doc.rkt

```diff
@@ -315,6 +316,44 @@
                                   #:end (Pos #:line line #:char really-indent))
                    #:newText new-text)]))
 
+(define (token-type-encoding token)
+  (index-of *semantic-token-types* (SemanticToken-type token)))
+
+(define (token-modifier-encoding token)
+  (define indexes (indexes-where *semantic-token-modifiers*
+                                 (λ (m) (memq m (SemanticToken-modifiers token)))))
+  (for/sum ([index indexes])
+    (expt 2 index)))
+
+(define (token-encoding doc token prev-pos)
+  (define-values (line ch) (doc-line/ch doc (SemanticToken-start token)))
+  (define-values (prev-line prev-ch) (doc-line/ch doc prev-pos))
+  (define delta-line (- line prev-line))
+  (define delta-start
+    (if (= line prev-line)
+        (- ch prev-ch)
+        ch))
+  (define len (- (SemanticToken-end token) (SemanticToken-start token)))
+  (define type (token-type-encoding token))
+  (define modifier (token-modifier-encoding token))
+  (values delta-line delta-start len type modifier))
+
+(define (doc-range-tokens doc path pos-start pos-end)
+  (define tokens (collect-semantic-tokens (Doc-text doc) (uri->path path)))
+  (define tokens-in-range
+    (filter-not (λ (tok) (or (<= (SemanticToken-end tok) pos-start)
+                             (>= (SemanticToken-start tok) pos-end)))
+                tokens))
+  (for/fold ([result '()]
+             [prev-pos 0]
+             #:result (let ()
+                        (flatten (reverse result))))
```

Comment:
Is this `(let () X)` cannot be `X`?

Reply:
It should be `X`. I inserted a `println` to debug in there but forgot to restore it.